pkg/reconciler/git/secret_copy.go

```diff
@@ -0,0 +1,138 @@
+package git
+
+import (
+	"context"
+	"io/ioutil"
+	coreV1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s "k8s.io/client-go/kubernetes"
+	"net/http"
+)
+
+type SecretFrom interface {
+	Get() (*coreV1.Secret, error)
+}
+
+type SecretCopy struct {
+	SecretFrom SecretFrom
+	SecretTo   SecretTo
+}
+
+func (c SecretCopy) Transfer() error {
+	secret, err := c.SecretFrom.Get()
+	if err != nil {
+		return err
+	}
+	err = c.SecretTo.createSecret(secret)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type SecretTo struct {
+	Key       string
+	Namespace string
+	Name      string
+
+	targetClientSet k8s.Interface
+}
+
+func NewFromURL(prefix string, targetClientSet k8s.Interface, configs map[string]string) *SecretCopy {
+	return &SecretCopy{
+		SecretFrom: &FromURL{
+			URL: configs[prefix+".secret.url"],
+			Key: configs[prefix+".secret.key"],
+		},
+		SecretTo: SecretTo{
+			Namespace:       configs[prefix+".secret.namespace"],
+			Name:            configs[prefix+".secret.name"],
+			Key:             configs[prefix+".secret.key"],
+			targetClientSet: targetClientSet,
+		},
+	}
+}
+
+func NewFromSecret(prefix string, targetClientSet k8s.Interface, inClusterClient k8s.Interface, configs map[string]string) *SecretCopy {
```

Comment:
What if we make a single constructor:
`func NewSecretCopy(prefix string, targetClientSet k8s.Interface) *SecretCopy` 

and two methods:
`func (s *SecretCopy) FromURL(configs map[string]string)` and 
`func (s *SecretCopy) FromSecret(inClusterClient k8s.Interface, configs map[string]string)`.

So it can be chained `NewSecretCopy(...).FromURL(...)`.

Anyway I don't like the fact that `configs` is map[string]string, since you don't know what should be inside. It looks more like dynamic typesetting from Python or JS, rather than static typesetting in Go.